### PR TITLE
write .npy files frame by frame

### PR DESCRIPTION
**Pull request by Bechir-Brahem**

### CODE REVIEW

File: pyctbgui/utils/numpyWriter/npy_writer.py [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+- 2 bytes (unsigned short) HEADER_LEN length of header to follow
+- Header as an ASCII dict terminated by \n padded with space \x20 to make sure
+we get len(magic string) + 2 + len(length) + HEADER_LEN divisible with 64
+Allocate enough space to allow for the data to grow
+"""
+
+import ast
+import logging
+import os
+import zipfile
```
Comment (erikfrojdh) [inline pyctbgui/utils/numpyWriter/npy_writer.py:18]: How is this used? Only valid for npz files?

File: pyctbgui/utils/numpyWriter/npy_writer.py [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+    headerLength = np.uint16(128)
+    FSEEK_FILE_END = 2
+    BUFFER_MAX = 500
+
+    def __init__(
+        self,
+        file: str | Path | zipfile.ZipExtFile,
+        frameShape: tuple = None,
+        dtype=None,
+        resetFile: bool = False,
```
Comment (erikfrojdh) [inline pyctbgui/utils/numpyWriter/npy_writer.py:38]: I would prefer to use a mode like open (https://docs.python.org/3/library/functions.html#open) which would be more intuitive for many people.
```
'r' = read default option
'w'=write, by default truncates existing files
'x'=exclusive creation for write, fails if file exists
```
```python
f = NumpyFileManager('somefile.npy') #opens
f = NumpyFileManager('somefile.npy', 'w') #for writing etc

```

File: pyctbgui/utils/numpyWriter/npy_writer.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+    ):
+        """
+        initiates a NumpyFileManager class for reading or writing bytes directly to/from a .npy file
+        @param file: path to the file to open or create
+        @param frameShape: shape of the frame ex: (5000,) for waveforms or (400,400) for image
+        @param dtype: type of the numpy array's header
+        @param resetFile: set to True if you wish to clear and write on an existing file
+        @param bufferMax: maximum length of the buffer that accumulates frames before flushing it  and calling syscalls
+        """
+        self.logger = logging.getLogger('npy_writer')
```
Comment (erikfrojdh) [inline pyctbgui/utils/numpyWriter/npy_writer.py:49]: To make it more widely usable I would remove the logging.

File: pyctbgui/utils/numpyWriter/npy_writer.py [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+            'shape': (self.frameCount, *self.frameShape)
+        }
+        np.lib.format.write_array_header_1_0(self.file, header_dict)
+
+    def writeOneFrame(self, frame: np.ndarray):
+        """
+        write one frame without buffering
+        @param frame: numpy array for a frame
+        """
+        assert frame.shape == self.frameShape
```
Comment (erikfrojdh) [inline pyctbgui/utils/numpyWriter/npy_writer.py:117]: Do you want to use a specific exception or are you happy with assertion error?

File: pyctbgui/utils/numpyWriter/npy_writer.py [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+            self.file.seek(0, self.FSEEK_FILE_END)
+            self.file.write(bytes(self.buffer))
+            self.buffer = bytearray()
+            self.updateHeader()
+            self.bufferCount = 0
+        if strict:
+            self.file.flush()
+            os.fsync(self.file)
+
+    def readFrames(self, frameStart: int, frameCount: int) -> np.ndarray:
```
Comment (erikfrojdh) [inline pyctbgui/utils/numpyWriter/npy_writer.py:157]: To me a more intuitive api would keep the position in the class 

```python
f = NumpyFileManager('somefile.py')
data = f.read(10)   #reads frames 0-10
data2 = f.read(3)   #reads frames 10-13
f.seek(20) #(or f.seek_frame(20)) move to frame 20
f.read(5) # reads frame 20-25

```

File: pyctbgui/utils/numpyWriter/npy_writer.py [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+        @param frame: numpy array for a frame
+        """
+        assert frame.shape == self.frameShape
+        assert frame.dtype == self.dtype
+        self.file.seek(0, self.FSEEK_FILE_END)
+        self.frameCount += 1
+        self.file.write(frame.tobytes())
+        self.updateHeader()
+
+    def addFrame(self, frame: np.ndarray):
```
Comment (erikfrojdh) [inline pyctbgui/utils/numpyWriter/npy_writer.py:124]: The addition of a buffer makes the class more complicated did you do performance tests to see that we need it or how did you reason about this feature. I think for our use case writing directly only could be ok... but I can be convinced.

**Review by erikfrojdh — COMMENTED** [review]
Added some comment's even though I see it's a draft PR. But this might save some time for you. Only looked at the npy_writer. 

Would it be possible to get the file writers in a PR that doesn't touch any other parts of the code?

File: pyctbgui/utils/numpyWriter/npy_writer.py (hunk 1 above)
Comment (Bechir-Brahem) [reply pyctbgui/utils/numpyWriter/npy_writer.py:18]: i have made a small usage.md file [link](https://github.com/slsdetectorgroup/pyctbgui/blob/feature/numpyWriter-Module/pyctbgui/utils/numpyWriter/usage.md) it explains what can this library do

**Pull request closed without merging**